generator/views/client_counts_view.py

```diff
@@ -0,0 +1,132 @@
+"""Class to describe a Client Counts View."""
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import Any, Dict, Iterator, List, Optional, Union
+
+from .view import View, ViewDict
+
+
+class ClientCountsView(View):
+    """A view for Client Counting measures."""
+
+    type: str = "client_counts_view"
+
+    default_dimension_groups: List[Dict[str, Union[str, List[str]]]] = [
+        {
+            "name": "since_first_seen",
+            "type": "duration",
+            "description": "Amount of time that has passed since the client was first seen.",
+            "sql_start": "CAST(${TABLE}.first_seen_date AS TIMESTAMP)",
+            "sql_end": "CAST(${TABLE}.submission_date AS TIMESTAMP)",
+            "intervals": ["day", "week", "month", "year"],
+        }
+    ]
+
+    default_dimensions: List[Dict[str, str]] = [
+        {
+            "name": "have_completed_period",
+            "type": "yesno",
+            "description": "Only for use with cohort analysis."
+            "Filter on true to remove the tail of incomplete data from cohorts."
+            "Indicates whether the cohort for this row have all had a chance to complete this interval."
+            "For example, new clients from yesterday have not all had a chance to send a ping for today.",
+            "sql": """
+              DATE_ADD(
+                {% if client_counts.first_seen_date._is_selected %}
+                  DATE_ADD(DATE(${client_counts.first_seen_date}), INTERVAL 1 DAY)
+                {% elsif client_counts.first_seen_week._is_selected %}
+                  DATE_ADD(DATE(${client_counts.first_seen_week}), INTERVAL 1 WEEK)
+                {% elsif client_counts.first_seen_month._is_selected %}
+                  DATE_ADD(PARSE_DATE('%Y-%m', ${client_counts.first_seen_month}), INTERVAL 1 MONTH)
+                {% elsif client_counts.first_seen_year._is_selected %}
+                  DATE_ADD(DATE(${client_counts.first_seen_year}, 1, 1), INTERVAL 1 YEAR)
+                {% endif %}
+                ,
+                {% if client_counts.days_since_first_seen._is_selected %}
+                  INTERVAL ${client_counts.days_since_first_seen} DAY
+                {% elsif client_counts.weeks_since_first_seen._is_selected %}
+                  INTERVAL ${client_counts.weeks_since_first_seen} WEEK
+                {% elsif client_counts.months_since_first_seen._is_selected %}
+                  INTERVAL ${client_counts.months_since_first_seen} MONTH
+                {% elsif client_counts.years_since_first_seen._is_selected %}
+                  INTERVAL ${client_counts.months_since_first_seen} YEAR
+                {% endif %}
+              ) < current_date
+              """,
+        }
+    ]
```

Comment:
I do wonder if python code is the best place to store these bits of lookml. We can figure this out at a later date of course.

Reply:
I think so, since we're storing them as native python objects. That big SQL string is a string in LookML as well, see https://github.com/mozilla/looker-spoke-default/blob/main/firefox_ios/views/client_counts.view.lkml#L23-L45 - that is, moving it to a separate representation would _still_ have to be a string.